---
title: "B1_ATAC_summary_Sp_RRBS"
author: "Marie Strader and Sam Bogan"
date: "6/27/2021"
output: github_document
---

This is an R markdown document detailing the wrangling, summary, and counting of ATAC-seq Tn5 inserts as measures of chromatin accessibility for the Sp_RRBS_ATAC repo, a documentation of analyses by Sam Bogan, Marie Strader, and Gretchen Hofmann that aimed to understand the gene regulatory effects of DNA methylation during transgenerational plasticity in the purple sea urchin *Strongylocentrotus purpuratus* and how these effects are regulated by other epigenomic and genomic states.

The code below wrangles a .gff3 file and formats it for use by ChIPSeeker using an R code chunk. It then uses a bash code chunk to concatenate 3 .bed files of Tn5 inserts from 39 hpf *S. purpuratus* larvae to be read into the proceeding R code chunk, which annotates and summarizes the Tn5 inserts. The final R code chunk creates files of genewise Tn5 insert counts across transcriptional start sites, firston introns, all introns, and all exons. 

This markdown outputs several files. The file 'ATAC_summary_peakSummaryALL.Rdata' is essential for proceeding to running B2_Integr_Sp_RRBS.Rmd, the R markdown file that integrates ATAC-seq, RRBS, and RNA-seq data.

ATAC-seq Tn5 insert .bed files are available under GEO experssion omnibus Bioproject PRJNA377768. ATAC-seq .bed files are large and must be imported to Sp_RRBS_ATAC/B_Integr_ATAC_RRBS_RNA/Input_data/ if this R markdown is to be run.

#Wrangle *S. purpuratus* annotations

```{r}

# Load dependencies
library( stringr )
library( plyr )
library( ChIPseeker )
library( GenomicFeatures )
library( GenomicRanges )
library( Repitools )

# Set path
setwd( "~/Documents/GitHub/Sp_RRBS_ATAC/B_Integr_ATAC_RRBS_RNA/Input_data/" )

# Load S. purpuratus GFF
spurp_txdb <- makeTxDbFromGFF( file = "Strongylocentrotus_purpuratus.Spur_3.1.42.gff3" ,
                               organism = "Strongylocentrotus purpuratus" )

#Partition S. purpuratus genomic features
promoter <- getPromoters( TxDb = spurp_txdb,
                          upstream = 3000,
                          downstream = 3000 )
introns <- getBioRegion( TxDb = spurp_txdb, by = "intron" )
exons <- getBioRegion( TxDb = spurp_txdb, by = "exon" )
genes <- getBioRegion( TxDb = spurp_txdb, by = "gene" )

# Save and
save( promoter, introns, exons, genes, file = "spurp_annots_txdb.Rdata" )
saveDb( spurp_txdb, file = "spurp_txdb.sqlite" )
ll = load( "Spurp_ATAC_prism.Rdata" )

spurp_txdb <- loadDb( file = "spurp_txdb.sqlite" )
ll2 = load( "spurp_annots_txdb.Rdata" )

```

#Concatenate Tn5 insert .bed files

```{bash}

#########################################################################################################################
##### Note: 39 hpf .bed files for ATAC-seq data must be imported to Sp_RRBS_ATAC/B_Integr_ATAC_RRBS_RNA/Input_data/ #####
#########################################################################################################################

# Ensure that you are in Input_data directory
cd ~/Documents/GitHub/Sp_RRBS_ATAC/B_Integr_ATAC_RRBS_RNA/Input_data/

### Concatenate Tn5 insert .bed files of 39 hpf S. purpuratus
## On the command line, cat the individual files:
# cat GSM2520650_spurp1_39hpf_MAPQ30.sorted.bed GSM2520651_spurp2_39hpf_MAPQ30.sorted.bed GSM2520652_spurp3_39hpf_MAPQ30.sorted.bed > Spurp_ATAC_39hpf_all_reps.bed

```

#Summarize annotation of Tn5 inserts

```{r} 

# Set path
setwd( "~/Documents/GitHub/Sp_RRBS_ATAC/B_Integr_ATAC_RRBS_RNA/Input_data/" )

# Read in concatenated .bed file
all_ATAC_reps <- readPeakFile( "Spurp_ATAC_39hpf_all_reps.bed" )
all_ATAC_reps_df <- data.frame( all_ATAC_reps )

# Create matrix of off tags relative to TSS
tagMatrixPM_all_reps <- getTagMatrix( all_ATAC_reps, windows = promoter )

# Heatmap of Tn5 inserts
tagHeatmap( tagMatrixPM_all_reps, xlim = c( -3000, 3000 ), color = "red" )

####Save this tag count plot and upload to Box####
Fig_2B <- plotAvgProf( tagMatrixPM_all_reps, xlim = c( -3000, 3000 ),
                    xlab = "Genomic Region (5'->3')", 
                    ylab = "Read Count Frequency", conf = TRUE )

# Print Figure 2B
Fig_2B

# Annotate peaks
peakAnnoM_all_reps <- annotatePeak( all_ATAC_reps, tssRegion = c( -3000, 3000 ),
                                    TxDb = spurp_txdb )

# Pie chart of Tn5 inserts by feature type
plotAnnoPie( peakAnnoM_all_reps )

# Save R data for genome txdb and peak annotation to output data directory
save( spurp_txdb, peakAnnoM_all_reps, 
      file = "~/Documents/GitHub/Sp_RRBS_ATAC/B_Integr_ATAC_RRBS_RNA/Output_data/Spurp_ATAC_prism.Rdata" )

# Save a GRanges object of Tn5 insert annotations - Store somewhere appropriate for large files
peakAnnoGR_all_reps <-as.GRanges( peakAnnoM_all_reps )
save( peakAnnoGR_all_reps, 
      file = "~/Documents/GitHub/Sp_RRBS_ATAC/B_Integr_ATAC_RRBS_RNA/peakAnnoGM_prism.Rdata" )

```

#Summarizing peak counts across genomic features

```{r}

ll = load( "peakAnnoGM_prism.Rdata" )
ll2 = load( "Spurp_ATAC_prism.Rdata" )

# Make peak annotation file into df for easier use
peakannotdf_all = data.frame( peakAnnoGR_all_reps )

# Get peaks for TSS, +/-500 from the TSS
TSS_Anno_all <- peakAnnoGR_all_reps[ abs( peakannotdf_all$distanceToTSS ) < 500 ]
TSS_Annodf_all = data.frame( TSS_Anno_all )

# Filter for unique 
TSSpeakGenes_all <- unique( TSS_Anno_all$transcriptId )
dim( TSSpeakGenes_all ) # How many genes with TSS Tn5 inserts?

# Sum counts per TSS by geneID
peakCovTSS_all <- count( TSS_Annodf_all$geneId )

## Need to fix geneIDs by converting common names to SPU_IDs
# Read in index of geneID types
annot = read.table( "gene_info_table_header.txt", header = T, sep = "\t", quote = NULL )
annot = as.data.frame( annot[ c( 2, 7) ], )
annot = as.data.frame( sapply( annot, toupper ) )

# Adjust column names of genewise TSS counts
names( peakCovTSS_all ) = c( "common_name", "peak_density_TSS" )

# Add spu_ids to TSS peak counts
TSS_Annodf2_all = merge( peakCovTSS_all, annot, by = "common_name", all.x = TRUE )

# Create intron1 and all introns peak summaries
peakannotdf_all$annotation <- sub( " ", "", peakannotdf_all$annotation )
peakannotdf_all$annotation <- as.factor( peakannotdf_all$annotation )
peakannotdf_all$feature <- as.factor( str_sub( peakannotdf_all$annotation, 1, 1 ) )

# Partition out peaks annotated within introns
peakIntron_all <- peakannotdf_all[ peakannotdf_all$feature == "I", ]
peakIntron_all$annotation <-sub ( " ","", peakIntron_all$annotation )

# Partition out peaks annotated within first introns
peakIntron1_all = peakIntron_all[ grep("intron 1 of", peakIntron_all$annotation ), ]

# How many genes have intron or first intron peaks?
length( unique( peakIntron1_all$geneId ) ) #X genes with peaks in intron 1
length( unique(peakIntron_all$geneId ) ) #X genes with peaks in introns

# Wrangle geneID variable and column IDs for first introns
peakIntron1_all$geneID = as.factor( peakIntron1_all$geneId )
peakCovIntron1_all = count( peakIntron1_all$geneId )
names( peakCovIntron1_all ) = c( "common_name", "peak_density_Intron1" )

# Wrangle geneID variable for introns
peakIntron_all$geneID = as.factor( peakIntron_all$geneId )

# Count intron peaks per gene
peakCovIntron_all = count( peakIntron_all$geneId )
names( peakCovIntron_all ) = c( "common_name", "peak_density_Introns" )

# Merge all intron count types into one count file w/ diff columsn for all and 1st introns
peakCovIntrons_all = merge( peakCovIntron_all, peakCovIntron1_all, by = "common_name", all = TRUE )

# Create exon peak summary
peakExon_all = peakannotdf_all[ peakannotdf_all$feature == "E", ]
peakExon_all$annotation <- sub( " ", "", peakExon_all$annotation )

# Wrangle geneID and column info
peakExon_all$geneID = as.factor( peakExon_all$geneId )

# Create genewise counts of peaks in exons
peakCovExon_all = count( peakExon_all$geneId )
names( peakCovExon_all ) = c( "common_name", "peak_density_Exons" )

# Save peak annotations and summaries
peaks_summary_all = merge( TSS_Annodf2_all, peakCovIntrons_all, by = "common_name", all = TRUE )
peaks_summary_all = merge( peaks_summary_all, peakCovExon_all, by = "common_name", all = TRUE )

# Save peakannotdf and peaks summary as R data files
save( peaks_summary_all, file = "ATAC_summary_peakSummaryALL.Rdata" )
save( peakannotdf_all, file = "ATAC_summary_peakAnnotDFALL.Rdata" )

```
